jbpm-workitems/jbpm-workitems-email/src/main/java/org/jbpm/process/workitem/email/TemplateManager.java

```diff
@@ -0,0 +1,201 @@
+import java.nio.file.WatchService;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import freemarker.cache.StringTemplateLoader;
+import freemarker.template.Configuration;
+import freemarker.template.Template;
+
+public class TemplateManager {
+    
+    private static final Logger logger = LoggerFactory.getLogger(TemplateManager.class);
+
+    private static TemplateManager INSTANCE = new TemplateManager();
+    
+    private StringTemplateLoader stringLoader = new StringTemplateLoader();
+    private Configuration cfg;
+    
+    private String templateDirectory = System.getProperty("org.jbpm.email.templates.dir");
+    private boolean watcherEnabled = Boolean.parseBoolean(System.getProperty("org.jbpm.email.templates.watcher.enabled", "false"));
+    
+    private TemplateDirectoryWatcher watcher;
+    
+    protected TemplateManager() {
+        cfg = new Configuration(Configuration.VERSION_2_3_26);
+        cfg.setTemplateLoader(stringLoader);
+        cfg.setDefaultEncoding("UTF-8");
+        
+        loadTemplates();
+        
+        if (watcherEnabled) {
+            this.watcher = new TemplateDirectoryWatcher(templateDirectory);
+            Thread watcherThread = new Thread(watcher, "Email Template Watch Thread");
+            watcherThread.start();
+        }
+    }
+    
+    public static TemplateManager get() {
+        return INSTANCE;
+    }
+    
+    public static synchronized TemplateManager reset() { 
+        // close in case there is watcher running to stop it
+        INSTANCE.close();
+        // create new instance
+        INSTANCE = new TemplateManager();
+        return INSTANCE;
+    }
+    
+    public String render(String templateName, Map<String, Object> parameters) {
+        StringWriter out = new StringWriter();
+        try {
+            Template template = cfg.getTemplate(templateName);
+            
+            template.process(parameters, out);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Template " + templateName + " not found", e);
+        }
+        return out.toString();
+    }
+    
+    protected void loadTemplates() {
+        if (templateDirectory != null) {
+            File directory = new File(templateDirectory);
+            
+            if (directory.exists() && directory.isDirectory()) {
+                File[] foundTemplates = directory.listFiles((dir, name) -> { return name.endsWith(".html");});
+                
+                for (File templateFile : foundTemplates) {
+                    
+                    loadTemplate(templateFile);
+                }
+            }
+        }
+    }
+    
+    protected void loadTemplate(File templateFile) {
+        String templateId = templateFile.getName().substring(0, templateFile.getName().lastIndexOf("."));
```

Comment:
This line could go into e.g. `resolveTemplateId` method, it will be reused in `removeTemplate` as well.

Reply:
done